clippy_lints/src/future_not_send.rs

```diff
@@ -0,0 +1,92 @@
+    /// functions and methods to implement the `Send` marker trait.
+    ///
+    /// **Why is this bad?** Returning a non-Send Future makes this Future
+    /// useless to code that uses multithreaded executors; even worse, the
+    /// error will typically not be reported in the source file where the
+    /// problem lies. Do not use this lint for single-threade (e.g. embedded)
+    /// targets.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// async fn not_send(bytes: Rc<[u8]>) {}
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// async fn is_send(bytes: Arc<[u8]>) {}
+    /// ```
+    pub FUTURE_NOT_SEND,
+    nursery,
+    "public Futures must be Send"
+}
+
+declare_lint_pass!(FutureNotSend => [FUTURE_NOT_SEND]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for FutureNotSend {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'a, 'tcx>,
+        kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        hir_id: HirId,
+    ) {
+        if let FnKind::Closure(_) = kind {
+            return;
+        }
+        let def_id = cx.tcx.hir().local_def_id(hir_id);
+        let fn_sig = cx.tcx.fn_sig(def_id);
+        let fn_sig = cx.tcx.erase_late_bound_regions(&fn_sig);
+        let ret_ty = fn_sig.output();
+        if let Opaque(id, subst) = ret_ty.kind {
+            let preds = cx.tcx.predicates_of(id).instantiate(cx.tcx, subst);
+            let mut is_future = false;
+            for p in preds.predicates {
+                if let Some(trait_ref) = p.to_opt_poly_trait_ref() {
+                    if Some(trait_ref.def_id()) == cx.tcx.lang_items().future_trait() {
+                        is_future = true;
+                        break;
+                    }
+                }
+            }
+            if is_future {
+                let send_trait = utils::get_trait_def_id(cx, &utils::paths::SEND_TRAIT).unwrap();
+                let span = decl.output.span();
+                let send_errors = cx.tcx.infer_ctxt().enter(|infcx| {
+                    let cause = traits::ObligationCause::misc(span, DUMMY_HIR_ID);
+                    let mut fulfillment_cx = traits::FulfillmentContext::new();
+                    fulfillment_cx.register_bound(&infcx, ParamEnv::empty(), ret_ty, send_trait, cause);
+                    if let Err(err) = fulfillment_cx.select_all_or_error(&infcx) {
+                        err
+                    } else {
+                        vec![]
+                    }
+                });
+                if !send_errors.is_empty() {
+                    utils::span_lint_and_then(cx, FUTURE_NOT_SEND, span, "returned Future is not Send", |db| {
+                        for FulfillmentError { obligation, .. } in send_errors.into_iter() {
+                            if let Some(trait_ref) = obligation.predicate.to_opt_poly_trait_ref() {
```

Comment:
You should just match on `obligation.predicate` to get `Predicate::Trait(trait_predicate, _)` and print `trait_predicate` instead, which should print as `T: Trait` instead of `<T as Trait>`, automatically.